src/UrlMatcher.php

```diff
@@ -0,0 +1,409 @@
+    }
+
+    private function marshalMethodNotAllowedResult(array $result): MatchingResult
+    {
+        $path = $result[1];
+
+        $allowedMethods = array_unique(
+            array_reduce(
+                $this->routeCollection->getRoutes(),
+                static function ($allowedMethods, Route $route) use ($path) {
+                    if ($path !== $route->getPattern()) {
+                        return $allowedMethods;
+                    }
+
+                    return array_merge($allowedMethods, $route->getMethods());
+                },
+                []
+            )
+        );
+
+        return MatchingResult::fromFailure($allowedMethods);
+    }
+
+    /**
+     * Inject routes into the underlying router
+     */
+    private function injectRoutes(): void
+    {
+        foreach ($this->routeCollection->getRoutes() as $index => $route) {
+            /** @var Route $route */
+            $this->fastRouteCollector->addRoute($route->getMethods(), $route->getPattern(), $route->getName());
+        }
+    }
+
+    /**
+     * Get the dispatch data either from cache or freshly generated by the
+     * FastRoute data generator.
+     *
+     * If caching is enabled, store the freshly generated data to file.
+     */
+    private function getDispatchData(): array
+    {
+        if ($this->hasCache) {
+            return $this->dispatchData;
+        }
+
+        $dispatchData = (array)$this->fastRouteCollector->getData();
+
+        if ($this->cacheEnabled) {
+            $this->cacheDispatchData($dispatchData);
+        }
+
+        return $dispatchData;
+    }
+
+    /**
+     * Load dispatch data from cache
+     * @throws \RuntimeException If the cache file contains invalid data
+     */
+    private function loadDispatchData(): void
+    {
+        set_error_handler(
+            static function () {
+            },
+            E_WARNING
+        ); // suppress php warnings
+        $dispatchData = include $this->cacheFile;
+        restore_error_handler();
+
+        // Cache file does not exist
+        if (false === $dispatchData) {
+            return;
+        }
+
+        if (!is_array($dispatchData)) {
+            throw new \RuntimeException(
+                sprintf(
+                    'Invalid cache file "%s"; cache file MUST return an array',
+                    $this->cacheFile
+                )
+            );
+        }
+
+        $this->hasCache = true;
+        $this->dispatchData = $dispatchData;
+    }
+
+    /**
+     * Save dispatch data to cache
+     * @param array $dispatchData
+     * @return int|false bytes written to file or false if error
+     * @throws \RuntimeException If the cache directory does not exist.
+     * @throws \RuntimeException If the cache directory is not writable.
+     * @throws \RuntimeException If the cache file exists but is not writable
+     */
+    private function cacheDispatchData(array $dispatchData)
```

Comment:
We are not interested in number of bytes written and are not checking for `false`. How about checking for `file_put_contents()` returns false and throwing exception? And the method then will return void.

Reply:
Done.